internal/storage/badger/store.go

```diff
@@ -271,38 +282,64 @@ func (bdb *badgerDB) CompareAndSet(key, expect, update []byte) (bool, error) {
 	return err == nil, err
 }
 
+const (
+	badgerSSTPrefix = "badger-snapshot-"
+)
+
 func (bdb *badgerDB) GetSnapshot() ([]byte, error) {
 	defer bdb.opts.statsCli.Timing("badger.snapshot.get.latency.ms", time.Now())
+
+	sstFile, err := storage.CreateTempFile(bdb.opts.sstDirectory, badgerSSTPrefix)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(sstFile.Name())
+
 	// TODO: Check if any options need to be set on stream
 	strm := bdb.db.NewStream()
-	snap := make(map[string][]byte)
+	w := bufio.NewWriter(sstFile)
+
 	strm.Send = func(list *badger_pb.KVList) error {
 		for _, kv := range list.Kv {
-			snap[string(kv.Key)] = kv.Value
+			entry := serverpb.PutRequest{Key: kv.Key, Value: kv.Value, ExpireTS: kv.ExpiresAt}
+			_, err := pbutil.WriteDelimited(w, &entry)
+			if err != nil {
+				return err
+			}
 		}
 		return nil
 	}
 	if err := strm.Orchestrate(context.Background()); err != nil {
 		return nil, err
 	}
 
-	var buf bytes.Buffer
-	err := gob.NewEncoder(&buf).Encode(snap)
-	return buf.Bytes(), err
+	w.Flush()
+	sstFile.Close()
+
+	return ioutil.ReadFile(sstFile.Name())
```

Comment:
Given that eventually all the contents of `sstFile` need to be read into memory, why can't we just use `bytes.Buffer` which implements `io.Writer` at the beginning ?

Reply:
I want to change `GetSnapshot` & `PutSnapshot` to use streaming io (from file) hence, the entire functions are written in streaming fashion, except this last line, which loads up the entire file in-memory. Once the snapshot side-car streaming changes are in place, this functions would return the `file` reference, which would just mean changing this line to return just the `sstFile`